pyccel/codegen/printing/ccode.py

```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
         Return
         ------
             String
-                Return a str that contains the declaration of a dummy data_buffer
-                       and a call to an operator which copies it to an NdArray struct
+                Return a str that contains the necessary declarations, allocations and copy operations
+                to create a NdArray
         """
+        self.add_import(c_imports['string'])
         rhs = expr.rhs
         lhs = expr.lhs
         if rhs.rank == 0:
             raise NotImplementedError(str(expr))
-        dummy_array_name = self.scope.get_new_name('array_dummy')
-        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
         arg = rhs.arg if isinstance(rhs, NumpyArray) else rhs
-
-        self.add_import(c_imports['string'])
+        lhs_address = self._print(ObjectAddress(lhs))
         if isinstance(arg, Variable):
-            arg = self._print(arg)
-            cpy_data = "memcpy({0}.{2}, {1}.{2}, {0}.buffer_size);\n".format(lhs, arg, dtype)
-            return '%s' % (cpy_data)
-        else :
-            if arg.rank > 1:
-                # flattening the args to use them in C initialization.
-                arg = self._flatten_list(arg)
-            arg = ', '.join(self._print(i) for i in arg)
-            dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
-            cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);\n".format(self._print(lhs), dummy_array_name, dtype)
-            return  '%s%s' % (dummy_array, cpy_data)
+            return f"array_copy_data({lhs_address}, {self._print(arg)}, 0);\n"
+        order = lhs.order
+        rhs_dtype = self._print(rhs.dtype)
+        declare_dtype = self.find_in_dtype_registry(rhs_dtype, rhs.precision)
+        dtype = self.find_in_ndarray_type_registry(rhs_dtype, rhs.precision)
+        flattened_list = self._flatten_list(arg)
+        i = 0
+        operations = ""
+        if order == "F":
+            nd = arg.rank
+            temp_array_name = self.scope.get_new_name('temp_array')
+            shape_dtype = self.find_in_dtype_registry('int', 8)
+            shape = ', '.join(self._print(elem) for elem in lhs.shape)
+            shape_Assign = "("+ shape_dtype +"[]){" + shape + "}"
+            temp_array_declaration = f"t_ndarray {temp_array_name} = {{.shape = NULL}};\n"
+            array_creation = f"{temp_array_name} = array_create({nd}, {shape_Assign}, {dtype}, {'false'}, {'order_c'});\n"
+            operations += temp_array_declaration + array_creation
+            copy_to = temp_array_name
```

Comment:
Although, I suggest to you figure out a way to do this in the semantic part instead of here because you'll have a variable declaration followed by allocation repeated too many times, especially if you repeat the call to this function.

Reply:
I disagree about moving it to the semantic stage. Things in the semantic stage are applied to all languages and there is no reason to have all this in a Fortran or python translation

Reply:
I'm also not sure how moving it to the semantic stage would avoid the problem you're describing

Reply:
True, I was just considering how the code is going to be printed. Instead of having a declaration followed by an allocation multiple times, I thought it might be cleaner to have bunch of declarations and then the allocations.

Reply:
Using `Declare` and `Allocate` does seem like a good idea (https://github.com/pyccel/pyccel/pull/1280/commits/17bec94d8ea5d3659999ec4105e96e6ae6b8c867)
@aihya @EmilyBourne I used `status="unallocated"` for the `Allocate` instance, but I could not understand what it refers to from the docs
(why do we have to possibility to set it to `allocated` when the class itself is used to allocate)

Reply:
The status is used in `_print_Allocate` to signify whether the variable should be deallocated if it already exist in the memory.

Reply:
> True, I was just considering how the code is going to be printed. Instead of having a declaration followed by an allocation multiple times, I thought it might be cleaner to have bunch of declarations and then the allocations.

I agree, but this can also be done from the `CCodePrinter`. You just need to insert the `Variable` into the scope so the `Declare` object is automatically created.

Reply:
> why do we have to possibility to set it to allocated when the class itself is used to allocate)

@aihya is right, consider the 2 following examples:
```python
a = np.empty(4)       # Allocate('a', allocated = 'unallocated')
....
a = np.empty(5)       # Allocate('a', allocated = 'allocated')
```
and
```python
for i in range(5):
    a = np.empty(i+1)    # Allocate('a', allocated = 'unknown') (unallocated when i == 0, allocated after that)
```